department-of-veterans-affairs/caseflow
# Craig/appeals 44958

Resolves [APPEALS-44958](https://jira.devops.va.gov/browse/APPEALS-44958)

# Description
- Change `appeal_affinities` table to allow null `affinity_start_date` values
- Add `after_save` hook to AppealAffinity model to update the appeal's DistributionTask instructions if the task is set to 'assigned'
- Add job UpdateAppealAffinityDatesJob
- Queue new job from the end of both a PushPriorityAppealsToJudgesJob and a StartDistributionJob
- Add `with_appeal_affinities` to DistributionScopes to easily join the `appeal_affinities` table during distribution
- Update the `appeals_ready_for_distribution` query used by the ACD testing dashboard to include an appeal's `affinity_start_date`
- Add new seed file for testing this work
- Disable unnecessary seed files temporarily on the feature branch
- Updated Appeal, Distribution, and Hearing factories
- Added DistributedCase factory
- Added Rspec tests for new job and `after_save` hook in AppealAffinity model

## Acceptance Criteria

## Testing Plan
On local or demo from a reset state or fresh deploy:
1. Clear all ready appeals by running these commands in console:
  - `DistributionTask.where(status: 'assigned')&.map(&:on_hold!)`
  - `VACOLS::Case.where(bfcurloc: ['81', '83'])&.map { |c| c.update!(bfcurloc: 'testing') }`
2. Run the new seed file in the rails console or via the below rake command (NOT in a rails console):
  - `bundle exec rake db:seed:appeals_for_affinity_calculation_updates`
3. Download the Ready to distribute appeals CSV from the ACD testing dashboard and verify that the affinity start dates match what the Veteran name describes
4. Run the new job from the rails console, providing the seeded distribution's ID as a parameter:
  - `distribution_id = Distribution.last`
  - `UpdateAppealAffinityDatesJob.perform_now(distribution_id)`
5. Downlaod the Ready to distribute appeals CSV again and verify that some affinity start dates have been updated and match with the Veteran name descriptions.
  - NOTE: No hearing nonpriority cases (AOD false, CAVC false) will be given an affinity start date by this seed data, and Veterans with the name "Vet Ready Shouldnt Get Affinity" should not be given an affinity start date


# Frontend
## User Facing Changes

 BEFORE|AFTER
 ---|---

## Storybook Story
*For Frontend (Presentation) Components*

# Backend
## Database Changes
*Only for Schema Changes*


## Integrations: Adding endpoints for external APIs

# Best practices
## Code Documentation Updates

## Tests
### Test Coverage
Did you include any test coverage for your code? Check below:

### Code Climate
Your code does not add any new code climate offenses? If so why?

## Monitoring, Logging, Auditing, Error, and Exception Handling Checklist

### Monitoring

### Logging

### Auditing

### Error Handling

### Exception Handling
